lib/chainable.ts

```diff
@@ -0,0 +1,218 @@
+  additional?: any;
+}
+
+/**
+ * Chainable function handler
+ *
+ * @export
+ * @class Chainable
+ * @template T
+ */
+export class Chainable<T> {
+  /**
+   * Call Stack
+   *
+   * @private
+   * @type {SingleCall[]}
+   * @memberof Chainable
+   */
+  private callQueue: SingleCall[] = [];
+
+  /**
+   * Creates an instance of Chainable.
+   * @param {T[]} input
+   * @memberof Chainable
+   */
+  constructor(public input: T[]) {}
+
+  /**
+   * aEvery method of Async-Ray lib
+   *
+   * @param {methods.CallBackEvery<T>} cb
+   * @returns {Chainable<T>}
+   * @memberof Chainable
+   */
+  aEvery(cb: methods.CallBackEvery<T>): Chainable<T> {
+    this.add(methods.aEvery, cb);
+
+    return this;
+  }
+
+  /**
+   * aFilter method of Async-Ray lib
+   *
+   * @param {methods.CallBackFilter<T>} cb
+   * @returns {Chainable<T>}
+   * @memberof Chainable
+   */
+  aFilter(cb: methods.CallBackFilter<T>): Chainable<T> {
+    this.add(methods.aFilter, cb);
+
+    return this;
+  }
+
+  /**
+   * aFind method of Async-Ray lib
+   *
+   * @param {methods.CallBackFind<T>} cb
+   * @returns {Chainable<T>}
+   * @memberof Chainable
+   */
+  aFind(cb: methods.CallBackFind<T>): Chainable<T> {
+    this.add(methods.aFind, cb);
+
+    return this;
+  }
+
+  /**
+   * aFindIndex method of Async-Ray lib
+   *
+   * @param {methods.CallBackFindIndex<T>} cb
+   * @returns {Chainable<T>}
+   * @memberof Chainable
+   */
+  aFindIndex(cb: methods.CallBackFindIndex<T>): Chainable<T> {
+    this.add(methods.aFindIndex, cb);
+
+    return this;
+  }
+
+  /**
+   * aMap method of Async-Ray lib
+   *
+   * @template R
+   * @param {methods.CallBackMap<T, R>} cb
+   * @returns {Chainable<T>}
+   * @memberof Chainable
+   */
+  aMap<R>(cb: methods.CallBackMap<T, R>): Chainable<T> {
+    this.add(methods.aMap, cb);
+
+    return this;
+  }
+
+  /**
+   * aReduce method of Async-Ray lib
+   *
+   * @template R
+   * @param {methods.CallBackReduce<T, R>} cb
+   * @returns {Chainable<T>}
+   * @memberof Chainable
+   */
+  aReduce<R>(cb: methods.CallBackReduce<T, R>, initialValue?: R): Chainable<T> {
+    this.add(methods.aReduce, cb, initialValue);
+
+    return this;
+  }
+
+  /**
+   * aReduceRight method of Async-Ray lib
+   *
+   * @template R
+   * @param {methods.CallBackReduceRight<T, R>} cb
+   * @returns {Chainable<T>}
+   * @memberof Chainable
+   */
+  aReduceRight<R>(
+    cb: methods.CallBackReduceRight<T, R>,
+    initialValue?: R
+  ): Chainable<T> {
+    this.add(methods.aReduceRight, cb, initialValue);
+
+    return this;
+  }
+
+  /**
+   * aSome method of Async-Ray lib
+   *
+   * @template R
+   * @param {methods.CallBackSome<T>} cb
+   * @returns {Chainable<T>}
+   * @memberof Chainable
+   */
+  aSome<R>(cb: methods.CallBackSome<T>): Chainable<T> {
+    this.add(methods.aSome, cb);
+
+    return this;
+  }
+
+  /**
+   * Process the call stack
+   *
+   * @returns {Promise<any>}
+   * @memberof Chainable
+   */
+  async process(): Promise<any> {
```

Comment:
I would favour an implementation without this method (at least without requirement to always call it - might be inconvenient and people may forget to call `process` and thus wonder what is wrong).
A flag (with default value) in the constructor could determine whether the call to `process` is needed or not. With that one would have the freedom do define the chain earlier and process it at a certain point by calling `process` explicitly. Defaultly the processing should immediately start.

Reply:
New implementation, does not require to call if the last method does not return an array.